### PR TITLE
Feature: reporting fn to allow more ergonomic use as a library

### PR DESCRIPTION
**Pull request by cap10morgan**
This is a follow-up to #333 with the implementation outlined by @weavejester over there.

I think the main decision I made in here was the name and signature of the "library entry point" fns. I went with `check-paths` and `fix-paths` that take a paths collection as a first arg and then an optional options map. I did this b/c the primary purpose of this refactor was to create library-friendly entry points that check / fix paths like the CLI tool entry points do but w/o doing things like printing to stdout or exiting the whole process.

But I don't have strong opinions about any of that, so let me know if you'd like to see any changes there or anywhere else.

### CODE REVIEW

**Comment by weavejester** [comment]
Thanks for the PR, and my apologies for not being clear about the design I was suggesting. In clojure.test and other similar libraries, the reporting function reacts to events like `:pass`, `:fail` and `:summary`, and then decides how it will output these. We could do something similar with cljfmt. Consider a multimethod:

```clojure
(defmulti cli-report
  (fn [event context data] event))
```

Used like:

```clojure
(defn check-no-config [{:keys [parallel? paths report] :as options}]
  (let [map*    (if parallel? pmap map)
        summary (->> paths
                     (mapcat (partial find-files options))
                     (map* (partial check-one options))
                     (reduce #(report :check/file %1 %2) options))]
    (report :check/summary summary nil)))
```

The reporter would be responsible for reporting data for the user. This includes both outputting on a file-by-file basis, and collating the data in a context map for use in the summary. We could use the options as the initial context, and add a `::count` key to it.

I'd imagine we'd want at least 4 events: `:check/file`, `:check/summary`, `:fix/file` and `:fix/summary`. Maybe a `:check/initial` and `:fix/initial` as well that would be no-op by default, but available for reporters that want to setup something like a progress bar.

**Comment by cap10morgan** [comment]
@weavejester Makes sense. I'll work on moving it in that direction next chance I get. Thanks for the feedback!

**Comment by cap10morgan** [comment]
@weavejester Any other TODO's here? Happy to refine if this needs some more work before merging.

**Comment by cap10morgan** [comment]
@weavejester just checking in again to see where we're at with this PR

**Comment by weavejester** [comment]
Apologies for the delay in looking at this. I'll review the code now.

File: cljfmt/src/cljfmt/tool.clj
Comment (weavejester) [inline cljfmt/src/cljfmt/tool.clj:123]: Can you write these methods a little more concisely? e.g.

```clojure
(defmethod report-cli :check/initial [_ _ _])
```

And:

```clojure
(defmethod report-lib :fix/file [_ _ data] data)
```

File: cljfmt/src/cljfmt/tool.clj
Comment (weavejester) [inline cljfmt/src/cljfmt/tool.clj:119]: These multimethods should explicitly define their signatures, e.g.

```clojure
(defmulti report-cli (fn [event _context _data] event))
```

Can you also group the `report-cli` and `report-lib` multimethods together?

File: cljfmt/src/cljfmt/tool.clj
Comment (weavejester) [inline cljfmt/src/cljfmt/tool.clj:104]: Prefer

```clojure
(defn- merge-statuses [a b]
```

There's no need to add in extra whitespace when it doesn't aid readability.

File: cljfmt/src/cljfmt/tool.clj
Comment (weavejester) [inline cljfmt/src/cljfmt/tool.clj:174]: Why add this function? Why not make it equivalent to using `report-lib` with `check`?

File: cljfmt/src/cljfmt/tool.clj
Comment (weavejester) [inline cljfmt/src/cljfmt/tool.clj:8]: Prefer `(:import [java.io File])`

File: cljfmt/src/cljfmt/tool.clj
Comment (weavejester) [inline cljfmt/src/cljfmt/tool.clj:103]: I don't understand the purpose of `merge-statuses`. Can you explain?

File: cljfmt/src/cljfmt/tool.clj
Comment (weavejester) [inline cljfmt/src/cljfmt/tool.clj:157]: The `report` should be a `reduce`, not a map. The reporting functions should be a type signature of:

```clojure
(report event context data) -> new-context
```

That way the type signature of the multi-method remains consistent. See the example in my earlier comment.

**Review by weavejester — CHANGES_REQUESTED** [review]
Thanks for the commits. I have some questions/suggestions. The primary issue to be solved is that the reporting function needs a consistent type signature.

File: cljfmt/src/cljfmt/tool.clj
Comment (cap10morgan) [reply cljfmt/src/cljfmt/tool.clj:174]: Well, the entire point of this PR is to make this more ergonomic to use as a library. So this is the library entry point fn for checking paths.

Not sure I understand this:
> Why not make it equivalent to using report-lib with check?

File: cljfmt/src/cljfmt/tool.clj
Comment (cap10morgan) [reply cljfmt/src/cljfmt/tool.clj:103]: It's the counterpart to `merge-counts` when you want a fuller report of what was done, which you often do when using this as a library.

File: cljfmt/src/cljfmt/tool.clj
Comment (weavejester) [reply cljfmt/src/cljfmt/tool.clj:174]: I mean something like:

```clojure
(check {:paths ["src"], :report report-lib})
```

File: cljfmt/src/cljfmt/tool.clj
Comment (cap10morgan) [reply cljfmt/src/cljfmt/tool.clj:157]: Yeah, I started with that. It's a little more complex and messy that way. Instead of just returning what is inherently a list of results, you have to stuff it into the original context somewhere, which starts out as the `options` map. So then the final return value is all of those options + wherever you stuffed the results.

I have a working refactor that puts them into a `:results` key. Happy to push that if you still prefer it.

**Comment by weavejester** [comment]
To give a little more context, this was the design I had in mind for the CLI reporter. Note that the method signature remains consistent: `event context data -> context`, with the initial context being the options map.

```clojure
(defmulti report-cli
  (fn [event _context _data] event))

(defmethod report-cli :check/init [_ context _] context)

(defmethod report-cli :fix/init [_ context _] context)

(defmethod report-cli :check/file [_ context file-status]
  (print-file-status context file-status)
  (update context :counts (fnil merge-counts {}) (:counts file-status)))

(defmethod report-cli :fix/file [_ context file-status]
  (print-file-status context file-status)
  context)

(defmethod report-cli :check/summary [_ {:keys [counts]} _]
  (print-final-count counts)
  (exit counts))

(defmethod report-cli :fix/summary [_ _ _])
```

**Comment by cap10morgan** [comment]
OK @weavejester I just pushed some changes that I think address your feedback

File: cljfmt/src/cljfmt/tool.clj
Comment (cap10morgan) [reply cljfmt/src/cljfmt/tool.clj:165]: I will also note that this is a big part of why I started with a different namespace. Naming this fn in here is tricky in terms of revealing its intention (and so was naming the other namespace as you probably recall; so that's perhaps not a great solution either).

But the idea is you would call this fn when using cljfmt as a library to check paths (instead of strings), and the existing fn(s) are still for internal CLI usage. I'm open to ideas about how to better name this, document that intention, etc.

But I do think it makes sense to have a "library entrypoint" fn that doesn't require setting options you'll want every time to make it be a well-behaved library (e.g. not calling `System/exit`, not printing to stdout, etc.).

File: cljfmt/src/cljfmt/tool.clj
Comment (weavejester) [inline cljfmt/src/cljfmt/tool.clj:208]: `doall` isn't required if there is a `reduce` before it.

Comment (cap10morgan) [reply cljfmt/src/cljfmt/tool.clj:208]: Ah, yep. Meant to revisit that. Thanks!

File: cljfmt/src/cljfmt/tool.clj
Comment (weavejester) [reply cljfmt/src/cljfmt/tool.clj:165]: I think I'd rather omit the `check-paths` and `fix-paths` wrapper functions for now. I don't think it's a huge deal to add an extra `:report` argument to cater for a relatively niche use-case.

I think it would also be a good idea to move `report-cli` to `cljfmt.report/console` and `report-lib` to `cljfmt.report/clojure`. However, don't feel you need to make that chage; I can always move things around after the PR is merged.

Comment (cap10morgan) [reply cljfmt/src/cljfmt/tool.clj:165]: OK, I can make both changes

File: cljfmt/src/cljfmt/tool.clj
Comment (weavejester) [inline cljfmt/src/cljfmt/tool.clj:115]: Maybe something more like:

```clojure
(defn- merge-statuses [results {:keys [file diff exception counts]}]
  (let [path (some-> file .getPath)]
    (-> results
        (update :counts merge-counts counts)
        (cond-> (and path (> (:incorrect counts) 1))
          (assoc-in [:incorrect-files path] {}))
        (cond-> (and path diff)
          (assoc-in [:incorrect-files path :diff] diff))
        (cond-> (and path exception)
          (assoc-in [:errored-files path :exception] exception)))))
```

File: cljfmt/src/cljfmt/tool.clj
Comment (cap10morgan) [reply cljfmt/src/cljfmt/tool.clj:165]: Would you prefer these changes as new, separate commits? Or should I just continue to amend the existing HEAD commit on this branch?

Comment (cap10morgan) [reply cljfmt/src/cljfmt/tool.clj:165]: I'll go ahead and push separate commits for easier review. Happy to squash later if desired.

Comment (cap10morgan) [reply cljfmt/src/cljfmt/tool.clj]: When I moved the reporting fns to the new `cljfmt.report` ns, it required one other change to avoid a circular dependency: The `project-path` fn is used by both `cljfmt.tool` fns and `cljfmt.report` fns, so it needed to be in a namespace they both required. `cljfmt.io` seemed like a natural fit, so I moved it there.

**Comment by cap10morgan** [comment]
@weavejester Should be ready for review again

File: cljfmt/src/cljfmt/tool.clj
Comment (weavejester) [inline cljfmt/src/cljfmt/tool.clj:75]: Is the merge of `{:report report/console}` necessary given the `:or` set in `check-no-config`?

And similarly for `fix` and `fix-no-config`?

**Review by weavejester — COMMENTED** [review]
Thanks for all your work! I just have one query this time.

File: cljfmt/src/cljfmt/tool.clj
Comment (cap10morgan) [reply cljfmt/src/cljfmt/tool.clj:75]: Good catch. No they are not! 😁 Will fix shortly

Comment (cap10morgan) [reply cljfmt/src/cljfmt/tool.clj]: Fixed in https://github.com/weavejester/cljfmt/pull/342/commits/ddf28c0b753ebd0fd51015f800c59b3133ec0972

**Comment by weavejester** [comment]
Thanks for your work on this. It all looks good. Can you squash your commits down and give it an appropriate commit message?

**Comment by cap10morgan** [comment]
> Thanks for your work on this. It all looks good. Can you squash your commits down and give it an appropriate commit message?

Done

**Comment by weavejester** [comment]
Thanks! Could you change the commit message to:

```
Add :report option to check and fix

Allows for a custom reporter function when checking/fixing files.
```

I think that's perhaps a little clearer.

**Comment by cap10morgan** [comment]
> Thanks! Could you change the commit message to:
> 
> ```
> Add :report option to check and fix
> 
> Allows for a custom reporter function when checking/fixing files.
> ```
> 
> I think that's perhaps a little clearer.

done

**Comment by cap10morgan** [comment]
@weavejester https://github.com/weavejester/cljfmt/pull/342/commits/69ca1078a91d5f6222b5feca3d32642a7f21d7bf fixes the CI issue. It was flagging that I hadn't wired up the `tool/exit` fn into the new event-based workflow. It should now mimic previous behavior. Let me know if that looks good and I'll re-squash.

**Comment by weavejester** [comment]
Thanks for the fix, and sorry for the delay again. It's been a busy week! It all looks good if you want to squash the changes.

**Comment by cap10morgan** [comment]
@weavejester done

**Comment by cap10morgan** [comment]
@weavejester Anything else needed here before merging?

**Comment by weavejester** [comment]
Sorry for the delay! I've been away for a couple of weeks. I don't think there's anything left to do, so I'll merge once the checks are done, but I may delay the release a little as I think about the format of the reporters.

**Pull request merged**
